TripalCultivate/TripalCultivate-Phenotypes
# G4.41 genus validator

**Issue #41 Genus Validator**

## Motivation



This PR adds another validation scope or level. The importer form will contain a genus select field that works in sync with the project autocomplete field in that it will autoselect the genus the project had been paired with.

![Screen Shot 2023-11-07 at 3 12 13 PM](https://github.com/TripalCultivate/TripalCultivate-Phenotypes/assets/15472253/aca6a0dc-55ef-497c-b030-401b1b212113)

## What does this PR do?
*Please describe each things this PR does. For example, a PR may 1) solve a specific bug, 2) create an auomated test to ensure it doesn't return.*

1. Upgrade the formValidate() of the importer method to support a sequence of validation scopes/levels.
2. Upgrade the formValidate() of the importer to allow to skip a validation scope or level and mark it as upcoming/todo. This is used when a prior scope failed and allow user to rectify the issue before proceeding.
3. Add a genus field that autoselects a genus based on the project entered.
4. Create a route to respond to fetch the genus of the project entered (AJAX request)
5. Upgrade the validator plugin system to allow additional properties to:
    a. headers - reference the headers defined in the importer. This for Header scope validation for later.
    b. skip - mark validator as upcoming or to do. 

## Testing
### Automated Testing
*Please describe each automated test this PR creates and provide a list of the assertions it makes using casual language.*
*Do not just say things like "asserts the array is not empty" but rather say "Ensures that the return value of method X with these parameters is not an empty array".*

This PR has automated test included to test the following:
1. Functional test - provide a project and set genus to the genus set for the project. This is an upgrade of the ImporterShareTest.php to test the new genus field added.
2. Kernel test - test genus validator plugin for:
Correct project-genus pair, incorrect genus, unpaired with a project to see if genus is active (Trait Importer where project is not required) and skip validator.
4. Kernel test - AJAX request controller to fetch genus of project.
Create a http request like in the form to fetch the genus of a project.

### Manual Testing
*Describe in detail how someone should manually test this functionality.*
*Make sure to include whether they need to build a docker from scratch, create any records, etc.*

1. Create test data
Create a project
Create a genus

2. Enable the module and run the Tripal Job created by the install.
    If the module had been already setup, go to configuration page 
    mysite/admin/tripal/extension/tripal-cultivate/phenotypes/ontology 
    
    and set the trait, method, unit and db configuration of the genus.

3. Pair the project - genus in project prop using the genus term in the configuration as the type, alternatively with the module properly configured use the service below.

```
\Drupal::service('trpcultivate_phenotypes.genus_project')
  ->setGenusToProject(PROJECT ID, 'GENUS', TRUE);
```

4. Load Tripal Cultivate PhenoShare Importer page.
5. Key in project and genus field should select the genus - Clicking the Validate Data File should proceed to next stage (all passed)
6. Key in project and genus field should select the genus but choose another genus - Clicking the Validate Data File should trigger a validation error, detailed in the validation result window
7. Key in non-existent project - Clicking the Validate Data File should trigger a project error and set the genus validation as to do/upcoming validation.
